hawkeye-api/src/backend.rs

```diff
@@ -0,0 +1,338 @@
+        .patch(
+            &templates::configmap_name(watcher.id.as_ref().unwrap()),
+            &patch_params,
+            &patch,
+        )
+        .await
+}
+
+pub async fn update_watcher_deployment(
+    k8s_client: &kube::Client,
+    watcher: &Watcher,
+) -> kube::Result<Deployment> {
+    // 2. Update Deployment with replicas=0
+    log::debug!("Updating Deployment instance");
+    let deploy = templates::build_deployment(
+        watcher.id.as_ref().unwrap(),
+        watcher.source.ingest_port,
+        watcher.tags.as_ref(),
+    );
+    let patch_params = PatchParams {
+        field_manager: Some(FIELD_MGR.to_owned()),
+        ..Default::default()
+    };
+    let patch = Patch::Merge(&deploy);
+    let deployments: Api<Deployment> = Api::namespaced(k8s_client.clone(), &config::NAMESPACE);
+    deployments
+        .patch(
+            &templates::deployment_name(watcher.id.as_ref().unwrap()),
+            &patch_params,
+            &patch,
+        )
+        .await
+}
+
+pub async fn update_watcher_service(
+    k8s_client: &kube::Client,
+    watcher: &Watcher,
+) -> kube::Result<Service> {
+    log::debug!("Updating Service instance");
+    let svc = templates::build_service(
+        watcher.id.as_ref().unwrap(),
+        watcher.source.ingest_port,
+        watcher.tags.as_ref(),
+    );
+    // TODO: Handle errors
+    let patch_params = PatchParams {
+        field_manager: Some(FIELD_MGR.to_owned()),
+        ..Default::default()
+    };
+    let patch = Patch::Merge(&svc);
+    let services: Api<Service> = Api::namespaced(k8s_client.clone(), &config::NAMESPACE);
+    services
+        .patch(
+            &templates::service_name(watcher.id.as_ref().unwrap()),
+            &patch_params,
+            &patch,
+        )
+        .await
+}
+
+// Apply Watcher updates and restart Watcher if it was previously running.
+pub async fn apply_watcher_updates(
+    k8s_client: &kube::Client,
+    watcher: &Watcher,
+) -> Result<(), kube::Error> {
+    // Update k8s resources.
+    let _ = update_watcher_configmap(k8s_client, watcher).await?;
+    let _ = update_watcher_deployment(k8s_client, watcher).await?;
+    let _ = update_watcher_service(k8s_client, watcher).await?;
+
+    // Only restart the worker if it was already started.
+    let retry_sleep = time::Duration::from_secs(1);
+    log::warn!("Watcher status: {:?}", watcher.status);
+    if watcher.status != Some(Status::Ready) {
+        let watcher_id = watcher.id.as_ref().unwrap();
+        let _ = stop_watcher(k8s_client, watcher_id).await?;
+
+        // k8s can take a few seconds to update, so retry a few times.
+        for _ in 1..8 {
+            let deployment_status = get_watcher_deployment(k8s_client, watcher_id).await?;
+            // .map_err(|_| warp::reject::custom(InternalError));
+            if deployment_status.get_watcher_status() == Status::Ready {
+                break;
+            }
+            let _ = tokio::time::sleep(retry_sleep);
```

Comment:
`tokio::time::sleep(retry_sleep)` returns a future, which needs to be awaited, otherwise it does nothing
```suggestion
            tokio::time::sleep(retry_sleep).await;
```